# Загрузка и исследование данных

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
#import seaborn as sn
from scipy import stats as st

In [2]:
geo_data = []
for i in range(3):
    geo_data_i = pd.read_csv('/datasets/geo_data_'+str(i)+'.csv')
    geo_data.append(geo_data_i)

for i in range(3):
    print(f'Анализируем данные по региону {i}:')
    geo_data[i].info()
    display(geo_data[i].head()) 
    display(geo_data[i].describe())
    print('Количество дубликатов:', geo_data[i].duplicated().sum())
    print('Количество скважин:', len(geo_data[i]['id'].unique()), 'из', geo_data[i].shape[0])
    print('===================================================================================')

Анализируем данные по региону 0:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

Количество дубликатов: 0
Количество скважин: 99990 из 100000
Анализируем данные по региону 1:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408

Количество дубликатов: 0
Количество скважин: 99996 из 100000
Анализируем данные по региону 2:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838

Количество дубликатов: 0
Количество скважин: 99996 из 100000


In [3]:
# разбиваем на признаки и целевые признаки
features = []
target = []
for i in range(3):
    print(f'Разбиваем на признаки и целевые признаки данные по региону {i}:')
    features.append(geo_data[i].drop(['id', 'product'], axis=1))
    features[i].info()
    target.append(geo_data[i]['product'])
    print('Количество записей в индексированном массиве целевых признаков:', len(target[i]))
    print('===================================================================================')

Разбиваем на признаки и целевые признаки данные по региону 0:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 3 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   f0      100000 non-null  float64
 1   f1      100000 non-null  float64
 2   f2      100000 non-null  float64
dtypes: float64(3)
memory usage: 2.3 MB
Количество записей в индексированном массиве целевых признаков: 100000
Разбиваем на признаки и целевые признаки данные по региону 1:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 3 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   f0      100000 non-null  float64
 1   f1      100000 non-null  float64
 2   f2      100000 non-null  float64
dtypes: float64(3)
memory usage: 2.3 MB
Количество записей в индексированном массиве целевых признаков: 100000
Разбиваем на признаки и целевые признаки данные по регио

### Выводы:
1. Пропусков нет.
2. Наименования столбцов корректны.
3. Типы данных в преобразовании не нуждаются.
4. Целевой признак содержится в столбце 'product'.
5. Для обучения модели значения столбца 'id' не представляют ценности.
6. По некоторым скважинам (до 10 в каждом регионе) почему-то есть разные данные не дублирующие друг-друга. Доля таких записей мала (менее 1%) и это можно не принимать во внимание. Хотя, странно еще и то, что id скважин в разных регионах совпадают (т.к. есть суммарно 49 повторов id в суммарном датасете, хотя ранее было 10+4+4=18).<br>

## Обучение и проверка модели

In [4]:
# Параметр рэндома
rs = np.random.RandomState(12345)

# Разделяю выборку на тренировочную и валидационную
features_train = []
features_valid = []
target_train = []
target_valid = []

for i in range (3):
    print(f'Разделяем выборку по региону {i} на тренировочную и валидационную:')
    features_train_i, features_valid_i, target_train_i, target_valid_i = train_test_split(features[i], target[i], test_size=0.25, random_state=rs)
    print(f'Размер тренировочной выборки в регионе {i}: {features_train_i.shape[0]} признаков и {len(target_train_i)} целевых признаков')
    print(f'Размер валидационной выборки в регионе {i}: {features_valid_i.shape[0]} признаков и {len(target_valid_i)} целевых признаков')
    features_train.append(features_train_i)
    features_valid.append(features_valid_i)
    target_train.append(target_train_i)
    target_valid.append(target_valid_i)
    print(type(target_valid_i))
    print(target_valid_i.index[0:10])
    print('===================================================================================')


Разделяем выборку по региону 0 на тренировочную и валидационную:
Размер тренировочной выборки в регионе 0: 75000 признаков и 75000 целевых признаков
Размер валидационной выборки в регионе 0: 25000 признаков и 25000 целевых признаков
<class 'pandas.core.series.Series'>
Int64Index([71751, 80493, 2655, 53233, 91141, 9539, 8482, 30284, 66393, 99584], dtype='int64')
Разделяем выборку по региону 1 на тренировочную и валидационную:
Размер тренировочной выборки в регионе 1: 75000 признаков и 75000 целевых признаков
Размер валидационной выборки в регионе 1: 25000 признаков и 25000 целевых признаков
<class 'pandas.core.series.Series'>
Int64Index([70156, 42796, 36723, 79446, 93328, 83694, 18295, 92055, 51870,
            67479],
           dtype='int64')
Разделяем выборку по региону 2 на тренировочную и валидационную:
Размер тренировочной выборки в регионе 2: 75000 признаков и 75000 целевых признаков
Размер валидационной выборки в регионе 2: 25000 признаков и 25000 целевых признаков
<class 'panda

In [5]:
model = []
predicted_valid = []

for i in range(3):
    model_i = LinearRegression()
    model_i.fit(features_train[i], target_train[i])
    predicted_valid_i = model_i.predict(features_valid[i])
    print(f'Качество модели по региону {i}: {np.round(model_i.score(features_valid[i], target_valid[i]), 3)}')
    model.append(model_i)
    predicted_valid.append(predicted_valid_i)
    print('Примеры предсказанных значений:')
    print(predicted_valid[i][0:5])
    print(type(predicted_valid[i]))    

Качество модели по региону 0: 0.28
Примеры предсказанных значений:
[95.89495185 77.57258261 77.89263965 90.17513418 70.51008829]
<class 'numpy.ndarray'>
Качество модели по региону 1: 1.0
Примеры предсказанных значений:
[ 54.23815201  83.21787849   2.97146723  25.91647996 111.34113673]
<class 'numpy.ndarray'>
Качество модели по региону 2: 0.203
Примеры предсказанных значений:
[112.56187784  69.46407333  69.86622946 110.33104829  95.12328758]
<class 'numpy.ndarray'>


In [6]:
for i in range (3):
    print(f'Средний запас предсказанного сырья в регионе {i}: {np.round(predicted_valid[i].mean(),3)}')
    print(f'Средний запас сырья в регионе {i} по правильным ответам: {np.round(target_valid[i].mean(), 3)}')
    mse_i = mean_squared_error(target_valid[i], predicted_valid[i])
    rmse_i = mse_i ** 0.5
    print(f'RMSE модели по региону {i}: {np.round(rmse_i, 3)}')
    print('')

Средний запас предсказанного сырья в регионе 0: 92.593
Средний запас сырья в регионе 0 по правильным ответам: 92.079
RMSE модели по региону 0: 37.579

Средний запас предсказанного сырья в регионе 1: 68.77
Средний запас сырья в регионе 1 по правильным ответам: 68.772
RMSE модели по региону 1: 0.89

Средний запас предсказанного сырья в регионе 2: 95.088
Средний запас сырья в регионе 2 по правильным ответам: 94.749
RMSE модели по региону 2: 39.958



### Анализ результатов:
1. Среднеквадратичная ошибка модели по регионам "0" и "2" довольно высокая, а по региону "1" она незначительна.
2. Средние по предсказанию и правильным ответам практически не различаются (есть небольшие отличия только по региону "2").<br>

## Подготовка к расчёту прибыли

In [7]:
# При разведке региона исследуют 500 точек, из которых с помощью машинного обучения выбирают 200 лучших для разработки.
POINTS_FOR_RESEARCH = 500
POINTS_FOR_DEVELOP = 200
# Бюджет на разработку скважин в регионе — 10 млрд рублей.
BUDGET_FOR_ONE_REGION = 10000000000
# При нынешних ценах один баррель сырья приносит 450 рублей дохода. Доход с каждой единицы продукта составляет 450 тыс. рублей, поскольку объём указан в тысячах баррелей.
REVENUE_FROM_PRODUCT = 450000
# После оценки рисков нужно оставить лишь те регионы, в которых вероятность убытков меньше 2.5%. Среди них выбирают регион с наибольшей средней прибылью.
LOSSS_PROBABILITY = 0.025

In [8]:
print('Средний запас сырья на 1 скважину достаточный для безубыточной разработки:', round(BUDGET_FOR_ONE_REGION / (POINTS_FOR_DEVELOP * REVENUE_FROM_PRODUCT))+1)

Средний запас сырья на 1 скважину достаточный для безубыточной разработки: 112


### Выводы по этапу подготовки расчёта прибыли:

1. Средний запас сырья на 1 скважину по регионам целом недостаточен для безубыточной разработки.
2. Но максимальные запасы сырья по отдельным скважинам в регионах 185, 137 и 190, а следовательно, возможно можно найти тот регион, где есть 200 (из 10000 в каждом) скважин, запасы которых дадут прибыль.
3. Необходимо понимать, что при среднем запасе на скважину  112 прибыль уйдет в ноль. Для более корректного порогового значения необходимо задать норму прибыли.
4. Скважины с запасом более 112 есть во всех регионах.

## Расчёт прибыли и рисков 

In [9]:
def profit_calculation (rrr, num):
    rrr = rrr.sort_values(by='predict', ascending=False).reset_index(drop=True) # Сортируем по предсказаниям
    rrr = rrr.head(num) # Взяли из них 200 лучших предсказаний
    sum_product = rrr['target'].sum() # Сумируем не предсказания, а исходные значения
    income = sum_product * REVENUE_FROM_PRODUCT
    revenue = income - BUDGET_FOR_ONE_REGION
    return revenue

In [10]:
%%time

NIT = 1000 # количество итераций бутстрепа
p = [[],[],[]] # Предсказанные профиты по регионам

for r in range(3): # Перебираем регионы
    print(f'Обрабатываем данные по региону {r}:')
    res = pd.DataFrame() # Создаем временный датафрейм
    res['target'] = target_valid[r]
    res['predict'] = predicted_valid[r]
    res.info()
        
    for i in range(NIT): # Итераци бутстрепа в каждом регионе
        rs = np.random.RandomState() # Для того, чтобы каждый раз брались случайные выборки
        res_sample = res.sample(POINTS_FOR_RESEARCH, replace=True, random_state=rs) # Взяли случайную выборку в 500 записей по скважинам
        profit = profit_calculation(res_sample, POINTS_FOR_DEVELOP)
        p[r].append(profit) # добавляем в список профитов по региону вычисленное значение по случайной выборке

    p[r] = pd.Series(p[r])
    print(f'Срeднее по прогнозируемой прибыли в регионе {r}: {round(p[r].mean())}')
    print(f'95%-ый доверительный интервал (по квантилям) по прогнозируемой прибыли в регионе {r}: {round(p[r].quantile(0.025))} - {round(p[r].quantile(0.975))}')
    print(f'Вероятность убытков в регионе {r}: {(p[r] < 0).mean()}%')
    print('===================================================================================')

Обрабатываем данные по региону 0:
<class 'pandas.core.frame.DataFrame'>
Int64Index: 25000 entries, 71751 to 43558
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   target   25000 non-null  float64
 1   predict  25000 non-null  float64
dtypes: float64(2)
memory usage: 585.9 KB
Срeднее по прогнозируемой прибыли в регионе 0: 389382758
95%-ый доверительный интервал (по квантилям) по прогнозируемой прибыли в регионе 0: -97338500 - 866460575
Вероятность убытков в регионе 0: 0.066%
Обрабатываем данные по региону 1:
<class 'pandas.core.frame.DataFrame'>
Int64Index: 25000 entries, 70156 to 93430
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   target   25000 non-null  float64
 1   predict  25000 non-null  float64
dtypes: float64(2)
memory usage: 585.9 KB
Срeднее по прогнозируемой прибыли в регионе 1: 462967014
95%-ый доверительный интервал (по квантилям) по прогнозируем

## Выводы:
Регион 1 по расчетам выше выглядит предпочтительным для разработки. У него наибольшая средняя прибыль при укладывающемся в заданный интервал риске убытков.